src/platform/nxp/k32w/k32w1/DiagnosticDataProviderImpl.cpp

```diff
@@ -0,0 +1,242 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *          Provides an implementation of the DiagnosticDataProvider object
+ *          for k32w1 platform.
+ */
+
+#include <platform/internal/CHIPDeviceLayerInternal.h>
+
+#include <crypto/CHIPCryptoPAL.h>
+#include <platform/DiagnosticDataProvider.h>
+#include <platform/nxp/k32w/k32w1/DiagnosticDataProviderImpl.h>
+
+#if CHIP_SYSTEM_CONFIG_USE_LWIP
+#include <lwip/tcpip.h>
+#endif
+
+extern "C" void xPortResetHeapMinimumEverFreeHeapSize(void);
+
+#include <openthread/platform/entropy.h>
+
+#include <platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.h>
+
+using namespace ::chip::app::Clusters::GeneralDiagnostics;
+
+namespace chip {
+namespace DeviceLayer {
+
+DiagnosticDataProviderImpl & DiagnosticDataProviderImpl::GetDefaultInstance()
+{
+    static DiagnosticDataProviderImpl sInstance;
+    return sInstance;
+}
+
+CHIP_ERROR DiagnosticDataProviderImpl::GetCurrentHeapFree(uint64_t & currentHeapFree)
+{
+    size_t freeHeapSize;
+
+    freeHeapSize    = xPortGetFreeHeapSize();
+    currentHeapFree = static_cast<uint64_t>(freeHeapSize);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DiagnosticDataProviderImpl::GetCurrentHeapUsed(uint64_t & currentHeapUsed)
+{
+    size_t freeHeapSize;
+    size_t usedHeapSize;
+
+    freeHeapSize = xPortGetFreeHeapSize();
+    usedHeapSize = MinimalHeapSize_c - freeHeapSize;
+
+    currentHeapUsed = static_cast<uint64_t>(usedHeapSize);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DiagnosticDataProviderImpl::GetCurrentHeapHighWatermark(uint64_t & currentHeapHighWatermark)
+{
+    size_t highWatermarkHeapSize;
+
+    highWatermarkHeapSize    = MinimalHeapSize_c - xPortGetMinimumEverFreeHeapSize();
+    currentHeapHighWatermark = static_cast<uint64_t>(highWatermarkHeapSize);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DiagnosticDataProviderImpl::ResetWatermarks()
+{
+    // If implemented, the server SHALL set the value of the CurrentHeapHighWatermark attribute to the
+    // value of the CurrentHeapUsed.
+
+    //    xPortResetHeapMinimumEverFreeHeapSize();
+
+    return CHIP_NO_ERROR;
```

Comment:
Shouldn't this return failure?

Reply:
Yes, I will remove the override and use the default implementation.